src/quantize_gptq.py

```diff
@@ -0,0 +1,125 @@
+import argparse
+import logging
+import random
+
+from transformers import AutoTokenizer
+import torch
+import numpy as np
+from datasets import load_dataset
+
+from auto_gptq import AutoGPTQForCausalLM, BaseQuantizeConfig
+
+logging.basicConfig(
+    format="%(asctime)s %(levelname)s [%(name)s] %(message)s",
+    level=logging.INFO,
+    datefmt="%Y-%m-%d %H:%M:%S",
+)
+
+
+def get_dataset(
+    nsamples, seed, seqlen, tokenizer, load_dataset_kwargs, buffer_size=5000
+):
+
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    np.random.seed(seed)
+    random.seed(seed)
+
+    # load raw dataset from Huggingface datasets
+    raw_data = load_dataset(**load_dataset_kwargs)
+
+    all_text = "\n\n".join(
+        raw_data["text"][:buffer_size]
+    )  # further reduce sample size (large enough to cover nsamples*seqlen tokens)
+
+    encoding = tokenizer(all_text, return_tensors="pt")
+
+    # gather dataset
+    dataset = []
+    for _ in range(nsamples):
+        i = random.randint(0, encoding.input_ids.shape[1] - seqlen - 1)
+        j = i + seqlen
+        inp = encoding.input_ids[:, i:j]
+        attention_mask = torch.ones_like(inp)
+        dataset.append({"input_ids": inp, "attention_mask": attention_mask})
+
+    return dataset
+
+
+def quantize(base_model, tokenizer_name=None, quantized_model_dir=None):
+    if tokenizer_name is None:
+        tokenizer_name = base_model
+
+    if quantized_model_dir is None:
+        quantized_model_dir = f"{base_model.split('/')[-1]}-GPTQ"
+
+    logging.info(f"base_model = {base_model}, tokenizer = {tokenizer_name}")
+
+    # base quantization config example # TODO. choose optimal one
+    quantize_config = BaseQuantizeConfig(
+        bits=4,  # quantize model to 4-bit
+        group_size=128,  # it is recommended to set the value to 128
+        desc_act=True,  # set to False can significantly speed up inference but the perplexity may slightly bad
+    )  # ppl is important than latency for now
+
+    # load un-quantized model, by default, the model will always be loaded into CPU memory
+    model = AutoGPTQForCausalLM.from_pretrained(base_model, quantize_config)
+
+    # get calibration data
+    # NOTE: please provide proper option for loading only subset of large datasets
+    load_dataset_kwargs = {
+        "path": "allenai/c4",
+        "split": "train",
+        "data_files": {"train": "en/c4-train.00000-of-01024.json.gz"},
+    }
+
+    try:
+        tokenizer = AutoTokenizer.from_pretrained(tokenizer_name, use_fast=False)
+    except Exception:
+        tokenizer = AutoTokenizer.from_pretrained(tokenizer_name, use_fast=True)
+
+    examples = get_dataset(
+        nsamples=128,
+        seed=0,
+        seqlen=2048,
+        tokenizer=tokenizer,
+        load_dataset_kwargs=load_dataset_kwargs,
+    )
```

Comment:
As we discussed before, please use [get_examples in src/dataset.py](https://github.com/Nota-NetsPresso/shortened-llm/blob/79e23b49fb7e60b46d68d66856e71bee4574e3f2/src/dataset.py#L90-L96). If needed, please modify the c4-load part in `src/dataset.py`.

Reply:
Thanks.

Reply:
Particularly, thanks for considering this part, which requires many modifications. Also, I appreciate the careful quality assessment for reproducibility!